src/write/xcoff.rs

```diff
@@ -0,0 +1,497 @@
+            StandardSection::TlsVariables => {
+                // Unsupported section.
+                (&[], &[], SectionKind::TlsVariables)
+            }
+            StandardSection::Common => {
+                // Unsupported section.
+                (&[], &[], SectionKind::Common)
+            }
+        }
+    }
+
+    pub(crate) fn xcoff_fixup_relocation(&mut self, mut relocation: &mut Relocation) -> i64 {
+        let constant = match relocation.kind {
+            RelocationKind::Relative => relocation.addend + 4,
+            _ => relocation.addend,
+        };
+        relocation.addend -= constant;
+        constant
+    }
+
+    pub(crate) fn xcoff_write(&self, buffer: &mut dyn WritableBuffer) -> Result<()> {
+        let is_64 = match self.architecture.address_size().unwrap() {
+            AddressSize::U8 | AddressSize::U16 | AddressSize::U32 => false,
+            AddressSize::U64 => true,
+        };
+
+        let (hdr_size, sechdr_size, rel_size, sym_size) = if is_64 {
+            (
+                mem::size_of::<xcoff::FileHeader64>(),
+                mem::size_of::<xcoff::SectionHeader64>(),
+                mem::size_of::<xcoff::Rel64>(),
+                mem::size_of::<xcoff::Symbol64>(),
+            )
+        } else {
+            (
+                mem::size_of::<xcoff::FileHeader32>(),
+                mem::size_of::<xcoff::SectionHeader32>(),
+                mem::size_of::<xcoff::Rel32>(),
+                mem::size_of::<xcoff::Symbol32>(),
+            )
+        };
+
+        // Calculate offsets and build strtab.
+        let mut offset = 0;
+        let mut strtab = StringTable::default();
+        // We place the shared address 0 immediately after the section header table.
+        let mut address = 0;
+
+        // XCOFF file header.
+        offset += hdr_size;
+        // Section headers.
+        offset += self.sections.len() * sechdr_size;
+
+        // Calculate size of section data.
+        let mut section_offsets = vec![SectionOffsets::default(); self.sections.len()];
+        for (index, section) in self.sections.iter().enumerate() {
+            let len = section.data.len();
+            let sectype = section.kind;
+            // Section address should be 0 for all sections except the .text, .data, and .bss sections.
+            if sectype == SectionKind::Data
+                || sectype == SectionKind::Text
+                || sectype == SectionKind::UninitializedData
+            {
+                section_offsets[index].address = address;
+            } else {
+                section_offsets[index].address = 0;
+            }
+            section_offsets[index].address = address;
+            if len != 0 {
+                // Set the default section alignment as 4.
+                offset = align(offset, 4);
+                section_offsets[index].data_offset = offset;
+                offset += len;
+                address += offset;
```

Comment:
I think this should only be updated if we're doing `section_offsets[index].address = address`.